ngraph/test/type_prop/proposal.cpp

```diff
@@ -143,6 +167,81 @@ TEST(type_prop, proposal_v0_shape_infer)
     ASSERT_EQ(op->get_output_shape(0), (Shape{batch_size * attrs.post_nms_topn, 5}));
 }
 
+TEST(type_prop, proposal_v0_dynamic_class_probs_dim1_batch_size_infer)
+{
+    op::ProposalAttrs attrs;
+    attrs.post_nms_topn = 1;
+    const size_t batch_size = 2;
+    auto class_probs =
+        make_shared<op::Parameter>(element::f32, PartialShape{Dimension::dynamic(), 2, 3, 4});
+    auto class_bbox_deltas = make_shared<op::Parameter>(element::f32, Shape{batch_size, 4, 3, 4});
+    auto image_shape = make_shared<op::Parameter>(element::f32, Shape{3});
+
+    auto op = make_shared<op::v0::Proposal>(class_probs, class_bbox_deltas, image_shape, attrs);
+    ASSERT_EQ(op->get_output_shape(0), (Shape{batch_size * attrs.post_nms_topn, 5}));
+}
+
+TEST(type_prop, proposal_v0_dynamic_bbox_deltas_dim1_batch_size_infer)
+{
+    op::ProposalAttrs attrs;
+    attrs.post_nms_topn = 1;
+    const size_t batch_size = 2;
+    auto class_probs =
+        make_shared<op::Parameter>(element::f32, PartialShape{Dimension::dynamic(), 2, 3, 4});
+    auto class_bbox_deltas = make_shared<op::Parameter>(element::f32, Shape{batch_size, 4, 3, 4});
+    auto image_shape = make_shared<op::Parameter>(element::f32, Shape{3});
+
+    auto op = make_shared<op::v0::Proposal>(class_probs, class_bbox_deltas, image_shape, attrs);
+    ASSERT_EQ(op->get_output_shape(0), (Shape{batch_size * attrs.post_nms_topn, 5}));
+}
+
+TEST(type_prop, proposal_v0_dynamic_class_probs_bbox_deltas_dim1_batch_size_infer)
+{
+    op::ProposalAttrs attrs;
+    auto class_probs =
+        make_shared<op::Parameter>(element::f32, PartialShape{Dimension::dynamic(), 2, 3, 4});
+    auto class_bbox_deltas =
+        make_shared<op::Parameter>(element::f32, PartialShape{Dimension::dynamic(), 4, 3, 4});
+    auto image_shape = make_shared<op::Parameter>(element::f32, Shape{3});
+
+    auto op = make_shared<op::v0::Proposal>(class_probs, class_bbox_deltas, image_shape, attrs);
+    ASSERT_EQ(op->get_output_partial_shape(0), (PartialShape{Dimension::dynamic(), 5}));
+}
+
+TEST(type_prop, proposal_v0_dynamic_image_shape_shape_infer)
+{
+    op::ProposalAttrs attrs;
+    attrs.base_size = 1;
+    attrs.pre_nms_topn = 20;
+    attrs.post_nms_topn = 200;
+    const size_t batch_size = 7;
+
+    auto class_probs = make_shared<op::Parameter>(element::f32, Shape{batch_size, 12, 34, 62});
+    auto class_bbox_deltas =
+        make_shared<op::Parameter>(element::f32, Shape{batch_size, 24, 34, 62});
+    auto image_shape = make_shared<op::Parameter>(element::f32, PartialShape::dynamic());
+    auto op = make_shared<op::v0::Proposal>(class_probs, class_bbox_deltas, image_shape, attrs);
+    ASSERT_EQ(op->get_output_shape(0), (Shape{batch_size * attrs.post_nms_topn, 5}));
+}
+
+TEST(type_prop, proposal_v0_everything_dynamic_shape_infer)
+{
+    op::ProposalAttrs attrs;
+    auto class_probs = make_shared<op::Parameter>(element::f32,
+                                                  PartialShape{Dimension::dynamic(),
```

Comment:
I think you could also write this PartialShape as `PartialShape::dynamic(4)`, where parameter for `dynamic()` indicates the rank.